hw/ip/otbn/doc/_index.md

```diff
@@ -0,0 +1,2514 @@
+
+```
+SUB <grd>, <grs1>, <grs2>
+```
+
+This instruction is defined in the RV32I instruction set.
+
+### AND
+
+**Bitwise AND.**
+
+```
+AND <grd>, <grs1>, <grs2>
+```
+
+This instruction is defined in the RV32I instruction set.
+
+### ANDI
+
+**Bitwise AND (immediate).**
+
+```
+ANDI <grd>, <grs1>, <imm>
+```
+
+This instruction is defined in the RV32I instruction set.
+
+### OR
+
+**Bitwise OR.**
+
+```
+OR <grd>, <grs1>, <grs2>
+```
+
+This instruction is defined in the RV32I instruction set.
+
+### ORI
+
+**Bitwise OR (immediate).**
+
+```
+ORI <grd>, <grs1>, <imm>
+```
+
+This instruction is defined in the RV32I instruction set.
+
+### XOR
+
+**Bitwise XOR.**
+
+```
+XOR <grd>, <grs1>, <rs2>
+```
+
+This instruction is defined in the RV32I instruction set.
+
+### XORI
+
+**Bitwise XOR (immediate).**
+
+```
+XORI <grd>, <grs1>, <imm>
+```
+
+This instruction is defined in the RV32I instruction set.
+
+### LW
+
+**Load Word.**
+
+```
+LW <grd>,<offset>(<grs1>)
+```
+
+This instruction is defined in the RV32I instruction set.
+
+### SW
+
+**Store Word.**
+
+```
+SW <grs2>, <offset>(<grs1>)
+```
+
+This instruction is defined in the RV32I instruction set.
+
+### BEQ
+
+**Branch Equal.**
+
+```
+BEQ <grs1>, <grs2>, <offset>
+```
+
+This instruction is defined in the RV32I instruction set.
+
+### BNE
+
+**Branch Not Equal.**
+
+```
+BNE <grs1>, <grs2>, <offset>
+```
+
+This instruction is defined in the RV32I instruction set.
+
+### LOOP
+
+**Loop (indirect).**
+Repeat a sequence of code multiple times.
+The number of iterations is a GPR value.
+The length of the loop is given as immediate.
+
+```
+LOOP <grs>, <bodysize>
+```
+
+Alternative assembly notation:
+The size of the loop body is given by the number of instructions in the parentheses.
+
+```
+LOOP <grs> (
+  # loop body
+)
+```
+
+<table>
+  <thead>
+    <tr><th>Assembly symbol</th><th>Description</th></tr>
+  </thead>
+  <tbody>
+    <tr>
+      <td><code>&lt;grs&gt;</code></td>
+      <td>Name of the GPR containing the number of iterations, encoded in the "grs" field.</td>
+    </tr>
+    <tr>
+      <td><code>&lt;bodysize&gt;</code></td>
+      <td>Number of instructions in the loop body (immediate).</td>
+    </tr>
+  </tbody>
+</table>
+
+### LOOPI
+
+**Loop Immediate.**
+Repeat a sequence of code multiple times.
+The number of iterations is given as an immediate, as is the length of the loop.
+
+```
+LOOPI <iterations>, <bodysize>
+```
+
+Alternative assembly notation. The size of the loop body is given by the number of instructions in the parentheses.
+
+```
+LOOPI <iterations> (
+  # loop body
+)
+```
+
+<table>
+  <thead>
+    <tr><th>Assembly symbol</th><th>Description</th></tr>
+  </thead>
+  <tbody>
+    <tr>
+      <td><code>&lt;iterations&gt;</code></td>
+      <td>Number of iterations (immediate).</td>
+    </tr>
+    <tr>
+      <td><code>&lt;bodysize&gt;</code></td>
+      <td>Number of instructions in the loop body (immediate).</td>
+    </tr>
+  </tbody>
+</table>
+
```

Comment:
We might want to specify that # of iterations can't be zero. For the LOOPI instruction it might be debatable whether we want to support this. From a pure algorithmic point of view, to not loose generality, it would be better to have it. This would mean an instant jump to the end of the loop in this case. My expectation, however, would be that this creates a non-negligible HW overhead.

Reply:
Good point, as it uncovers a potential misunderstanding.

I thought of the iteration count as "how often is the code repeated", and not "how often is it executed". Meaning: the code in the loop body is always executed at least once.

If "loop count" is the "number of repeats", then setting it to 0 would mean execute the code coming next, but don't jump back to the beginning after the loop body ends.

Your suggestion to disallow 0 and my interpretation are probably identical, except for wasting one value in the immediate if we disallow "loop count == 0". I'd therefore change the spec to read something like:


```
LOOPI <repetitions>, <bodysize>
```

repetitions: Number of times the loop body is repeated.
repetition == 0 means the loop body is executed only once,  repetition == 1 means the loop body is executed twice, etc.